dissect/cstruct/expression.py

```diff
@@ -1,84 +1,298 @@
 from __future__ import annotations
 
-from typing import TYPE_CHECKING, Dict
+import string
+from typing import TYPE_CHECKING, Callable, Optional, Union
+
+from dissect.cstruct.exceptions import ExpressionParserError, ExpressionTokenizerError
 
 if TYPE_CHECKING:
     from dissect.cstruct import cstruct
 
 
+class ExpressionTokenizer:
+    def __init__(self, expression: str):
+        self.expression = expression
+        self.pos = 0
+        self.tokens = []
+
+    def equal(self, token: str, expected: Union[str, str[str]]) -> bool:
+        if isinstance(expected, set):
+            return token in expected
+        else:
+            return token == expected
+
+    def alnum(self, token: str) -> bool:
+        return token.isalnum()
+
+    def alpha(self, token: str) -> bool:
+        return token.isalpha()
+
+    def digit(self, token: str) -> bool:
+        return token.isdigit()
+
+    def hexdigit(self, token: str) -> bool:
+        return token in string.hexdigits
+
+    def operator(self, token: str) -> bool:
+        return token in {"*", "/", "+", "-", "%", "&", "^", "|", "(", ")", "~"}
+
+    def match(
+        self,
+        func: Optional[Callable[[str], bool]] = None,
+        expected: Optional[str] = None,
+        consume: bool = True,
+        append: bool = True,
+    ) -> bool:
+        if self.eol():
+            return False
+
+        token = self.get_token()
+
+        if expected and self.equal(token, expected):
+            if append:
+                self.tokens.append(token)
+            if consume:
+                self.consume()
+            return True
+
+        if func and func(token):
+            if append:
+                self.tokens.append(token)
+            if consume:
+                self.consume()
+            return True
+
+        return False
+
+    def consume(self) -> None:
+        self.pos += 1
+
+    def eol(self) -> bool:
+        return self.pos >= len(self.expression)
+
+    def get_token(self) -> str:
+        if self.eol():
+            raise ExpressionTokenizerError(f"Out of bounds index: {self.pos}, length: {len(self.expression)}")
+        return self.expression[self.pos]
+
+    def tokenize(self) -> list[str]:
+        token = ""
+
+        # Loop over expression runs in linear time
+        while not self.eol():
+            # If token is a single character operand add it to tokens
+            if self.match(self.operator):
+                continue
+
+            # If token is a single digit, keep looping over expression and build the number
+            elif self.match(self.digit, consume=False, append=False):
+                token += self.get_token()
+                self.consume()
+
+                # Support for binary and hexadecimal notation
+                if self.match(expected={"x", "X", "b", "B"}, consume=False, append=False):
```

Comment:
This breaks on octal notation, 0o777 etc.. Maybe put this dict in a const also, otherwise you have to update it in multiple places.

Reply:
This does not seem to be valid C?
```
x.c:4:11: error: invalid suffix 'o7' on integer constant
        int a = 0o7;
```

Reply:
> This breaks on octal notation, 0o777 etc.. Maybe put this dict in a const also, otherwise you have to update it in multiple places.

0o777 is not valid C. We parse an octal number in C and convert it to a Python-style octal notation.

```
                if len(token) > 1 and token[0] == "0" and token[1] not in ("x", "X", "b", "B"):
                    token = token[:1] + "o" + token[1:]
```
I hope this adds some clarity unless I misunderstood something.

Reply:
Ah ok, my bet, I thought this expression parser allowed Python-style expressions, but it makes sense to stick to C.

Reply:
Does that also address your other comments?

Reply:
Yes, thanks. I am currently test-driving this branch while working on things. So far it seems to work well.

Reply:
Alright, I've marked your other comments as resolved. Let me know once you're satisfied!